test/javascripts/turbolinks_test.coffee

```diff
@@ -1,210 +1,946 @@
+          $(document).on 'page:after-link-inserted', linkTagInserted
+          visit url: 'fourLinksInHeadCDAB', ->
+            expectedLinkOrder = ['c.css', 'd.css', 'a.css', 'b.css']
+            expectedLinkOrder.forEach (link, index) ->
+              assert.equal(index, linkOrderPosition(link), link + ' should be in position ' + index)
+            assert.equal(linkTagInserted.callCount, 2) # Move A,B or C,D
+            done()
+
+      it 'maintains order when moving from a head with links ABCD to links CDBA.', (done) ->
+        visit url: 'fourLinksInHeadABCD', ->
+          linkTagInserted = sinon.spy()
+          $(document).on 'page:after-link-inserted', linkTagInserted
+          visit url: 'fourLinksInHeadCDBA', ->
+            expectedLinkOrder = ['c.css', 'd.css', 'b.css', 'a.css']
+            expectedLinkOrder.forEach (link, index) ->
+              assert.equal(index, linkOrderPosition(link), link + ' should be in position ' + index)
+            assert.equal(linkTagInserted.callCount, 2) # Move A,B
+            done()
+
+      it 'maintains order when moving from a head with links ABCD to links CBDA.', (done) ->
+        visit url: 'fourLinksInHeadABCD', ->
+          linkTagInserted = sinon.spy()
+          $(document).on 'page:after-link-inserted', linkTagInserted
+          visit url: 'fourLinksInHeadCBDA', ->
+            expectedLinkOrder = ['c.css', 'b.css', 'd.css', 'a.css']
+            expectedLinkOrder.forEach (link, index) ->
+              assert.equal(index, linkOrderPosition(link), link + ' should be in position ' + index)
+            assert.equal(linkTagInserted.callCount, 2) # Move A,B or A,C
+            done()
+
+      it 'maintains order when moving from a head with links ABCD to links DCBA.', (done) ->
+        visit url: 'fourLinksInHeadABCD', ->
+          linkTagInserted = sinon.spy()
+          $(document).on 'page:after-link-inserted', linkTagInserted
+          visit url: 'fourLinksInHeadDCBA', ->
+            expectedLinkOrder = ['d.css', 'c.css', 'b.css', 'a.css']
+            expectedLinkOrder.forEach (link, index) ->
+              assert.equal(index, linkOrderPosition(link), link + ' should be in position ' + index)
+            assert.equal(linkTagInserted.callCount, 3) # Move any 3
+            done()
+
+  describe 'head script tag tracking', ->
+    it 'dispatches page:after-script-inserted event when inserting a script on navigation', (done) ->
+      scriptTagInserted = sinon.spy()
+      $(document).on 'page:after-script-inserted', scriptTagInserted
+
+      visit url: 'noScriptsOrLinkInHead', ->
+        visit url: 'singleScriptInHead', ->
+          assert.equal(scriptTagInserted.callCount, 1)
+          done()
+
+    it 'inserts script with a new src into empty head on navigation', (done) ->
+      visit url: 'noScriptsOrLinkInHead', ->
+        visit url: 'singleScriptInHead', ->
+          assert(hasScriptOnce('foo.js'), 'Should insert foo.js tag only once.')
```

Comment:
Can't put my finger on it, but this assertion pattern seems funky.  There's never a scenario in which we'd want a duplicate script, so it feels like this should be called `hasScript`, and just fail/throw if it finds one.

When flipping from here to test data and back, it's really difficult to maintain a mental model of before/after.  Consider a more explicit pattern like:

```coffeescript
visit url: 'noScriptsOrLinkInHeader', ->
  assertScripts([])
  visit url: 'singleScriptInHeader', ->
    assertScripts(['foo.js'])
```

Custom assertions tend to be confusing, but they're maybe worth the hassle in this case.